lib/scheduler.ml

```diff
@@ -0,0 +1,198 @@
+    mutable pipe_out: Unix.file_descr option;
+    mutable pipe_in: Unix.file_descr option;
+    (* Indicates that a signal arrived before a wait: *)
+    mutable signalled: bool;
+    m: Mutex.t
+  }
+
+  let make () =
+    { pipe_out = None;
+      pipe_in = None;
+      signalled = false;
+      m = Mutex.create () }
+
+  exception Pre_signalled
+
+  let wait (x: t) (seconds: float) =
+    let to_close = ref [ ] in
+    let close' fd =
+      if List.mem fd !to_close then Unix.close fd;
+      to_close := List.filter (fun x -> fd <> x) !to_close in
+    finally
+      (fun () ->
+         try
+           let pipe_out = mutex_execute x.m
+               (fun () ->
+                  if x.signalled then begin
+                    x.signalled <- false;
+                    raise Pre_signalled;
+                  end;
+                  let pipe_out, pipe_in = Unix.pipe () in
+                  (* these will be unconditionally closed on exit *)
+                  to_close := [ pipe_out; pipe_in ];
+                  x.pipe_out <- Some pipe_out;
+                  x.pipe_in <- Some pipe_in;
+                  x.signalled <- false;
+                  pipe_out) in
+           let r, _, _ = Unix.select [ pipe_out ] [] [] seconds in
+           (* flush the single byte from the pipe *)
+           if r <> [] then ignore(Unix.read pipe_out (String.create 1) 0 1);
+           (* return true if we waited the full length of time, false if we were woken *)
+           r = []
+         with Pre_signalled -> false
+      )
+      (fun () ->
+         mutex_execute x.m
+           (fun () ->
+              x.pipe_out <- None;
+              x.pipe_in <- None;
+              List.iter close' !to_close)
+      )
+
+  let signal (x: t) =
+    mutex_execute x.m
+      (fun () ->
+         match x.pipe_in with
+         | Some fd -> ignore(Unix.write fd "X" 0 1)
+         | None -> x.signalled <- true 	 (* If the wait hasn't happened yet then store up the signal *)
+      )
+end
+
+type item = {
+  id: int;
+  name: string;
+  fn: unit -> unit
+}
+
+let schedule = ref Int64Map.empty
+let delay = Delay.make ()
+let next_id = ref 0
+let m = Mutex.create ()
+
+type time =
+  | Absolute of int64
+  | Delta of int [@@deriving rpc]
+
+type t = int64 * int [@@deriving rpc]
+
+let now () = Unix.gettimeofday () |> ceil |> Int64.of_float
+
+module Dump = struct
+  type u = {
+    time: int64;
+    thing: string;
+  } [@@deriving rpc]
+  type t = u list [@@deriving rpc]
+  let make () =
+    let now = now () in
+    mutex_execute m
+      (fun () ->
+         Int64Map.fold (fun time xs acc -> List.map (fun i -> { time = Int64.sub time now; thing = i.name }) xs @ acc) !schedule []
+      )
+end
+
+let one_shot time (name: string) f =
+  let time = match time with
+    | Absolute x -> x
+    | Delta x -> Int64.(add (of_int x) (now ())) in
+  let id = mutex_execute m
+      (fun () ->
+         let existing =
+           if Int64Map.mem time !schedule
```

Comment:
As above, I usually like more 
```ocaml
try Some (Int64Map.find time !schedule) with Not_found -> []
```
(same at L158). But I am not sure what is the idiomatic choice.